### DISCUSSION TITLE
How to respond with `InitializeError` on Initialize

### DISCUSSION THREAD

**Discussion by Badhi** [openlawlibrary/pygls#512] [Q&A]
I want to override the behavior for `Initialize` feature on the server and was able to do it overriding the `lsp_initialize` of `LanguageProtocol` like below

```python
class CustomeProtocol(LanguageProtocol):
     def lsp_initialize(config):
         result = super().lsp_initialize(config)
         status = my_server_initialize(config)
         return result
```

In the protocol I could see that we could use `InitializeError` to update client on server initialization failure. But I can't find a way to return this error along with the results. Is there a way to do it?

**Comment by tombh** [comment]
Currently Pygls' return signature for `LanguageProtocol.lsp_initialize` is just `InitializeResult`. But I think it should actually be `Union[InitializeResult, InitializeError]`. So strictly speaking, according to the spec, you can't return both an `InitializeResult` and an `InitializeError` at the same time to the client. And also the `InitializeError` interface only has one boolean field for whether the client should retry or not: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initializeError So I don't think you can return any more details with the error.

Either way I think you could return the `InitializeError` like this:

```python
from lsprotocol import types

class CustomeProtocol(LanguageProtocol):
     def lsp_initialize(config):
         ...
         return types.InitializeError(retry=False)
```

Or maybe you could let the server start anyway, store the error, and report it with `server.report_server_error()` once the server is up and running?
